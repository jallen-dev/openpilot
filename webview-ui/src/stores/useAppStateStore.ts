import { create } from "zustand"

interface AppState {
  workspaceIndexed: boolean
  setWorkspaceIndexed: (workspaceIndexed: boolean) => void
}

export const useAppStateStore = create<AppState>()((set, get) => ({
  workspaceIndexed: false,
  setWorkspaceIndexed: (workspaceIndexed: boolean) => {
    set({ workspaceIndexed })
  }
}))
