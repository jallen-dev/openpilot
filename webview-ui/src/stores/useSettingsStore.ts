import { create } from "zustand"

import { ApiKeys } from "../../../src/shared/ApiKeys"
import { Configuration } from "../../../src/shared/Configuration"
import { Model } from "../../../src/shared/Model"

interface SettingsState {
  model: Model
  apiKeys: ApiKeys
  includeFiles: boolean
  produceDiffs: boolean
  setModel: (model: Model) => void
  setApiKeys: (apiKeys: ApiKeys) => void
  setIncludeFiles: (includeFiles: boolean) => void
  setProduceDiffs: (produceDiffs: boolean) => void
  setConfiguration: (configuration: Configuration) => void
}

export const useSettingsStore = create<SettingsState>()((set, get) => ({
  model: "gpt-3.5-turbo",
  apiKeys: {},
  includeFiles: false,
  produceDiffs: false,
  setModel: (newModel) => set({ model: newModel }),
  setApiKeys: (newKeys) => {
    const { apiKeys } = get()
    set({ apiKeys: { ...apiKeys, ...newKeys } })
  },
  setIncludeFiles: (includeFiles) => set({ includeFiles }),
  setProduceDiffs: (produceDiffs) => set({ produceDiffs }),
  setConfiguration: (configuration: Configuration) => {
    const { includeFiles, produceDiffs } = configuration
    set({ includeFiles, produceDiffs })
  }
}))
