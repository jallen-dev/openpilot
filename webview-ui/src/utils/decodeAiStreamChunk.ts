const decoder = new TextDecoder()
export function decodeAIStreamChunk(chunk: Uint8Array): string {
  return decoder.decode(chunk)
}
