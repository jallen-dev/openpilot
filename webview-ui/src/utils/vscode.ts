import type { WebviewApi } from "vscode-webview"

import { WebviewMessage } from "../../../src/shared/OpenpilotMessage"
import { AppState } from "../types/AppState"

/**
 * A utility wrapper around the acquireVsCodeApi() function, which enables
 * message passing and state management between the webview and extension
 * contexts.
 *
 * This utility also enables webview code to be run in a web browser-based
 * dev server by using native web browser features that mock the functionality
 * enabled by acquireVsCodeApi.
 */
class VSCodeAPIWrapper {
  private readonly vsCodeApi: WebviewApi<unknown> | undefined

  constructor() {
    // Check if the acquireVsCodeApi function exists in the current development
    // context (i.e. VS Code development window or web browser)
    if (typeof acquireVsCodeApi === "function") {
      this.vsCodeApi = acquireVsCodeApi()
    }
  }

  /**
   * Post a message (i.e. send arbitrary data) to the owner of the webview.
   *
   * @remarks When running webview code inside a web browser, postMessage will instead
   * log the given message to the console.
   *
   * @param message Abitrary data (must be JSON serializable) to send to the extension context.
   */
  public postMessage(message: WebviewMessage) {
    if (this.vsCodeApi) {
      this.vsCodeApi.postMessage(message)
    } else {
      console.log(message)
    }
  }

  /**
   * Get the persistent state stored for this webview.
   *
   * @remarks When running webview source code inside a web browser, getState will retrieve state
   * from local storage (https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).
   *
   * @return The current state or `undefined` if no state has been set.
   */
  public getState() {
    if (this.vsCodeApi) {
      return this.vsCodeApi.getState() as AppState
    } else {
      const state = localStorage.getItem("vscodeState")
      return state ? (JSON.parse(state) as AppState) : undefined
    }
  }

  /**
   * Set the persistent state stored for this webview.
   *
   * @remarks When running webview source code inside a web browser, setState will set the given
   * state using local storage (https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).
   *
   * @param newState New persisted state. This must be a JSON serializable object. Can be retrieved
   * using {@link getState}.
   *
   * @return The new state.
   */
  public setState<T extends Partial<AppState> | undefined>(newState: T): T {
    const state = this.getState() ?? {}
    if (this.vsCodeApi) {
      return this.vsCodeApi.setState({ ...state, ...newState })
    } else {
      localStorage.setItem(
        "vscodeState",
        JSON.stringify({ ...state, ...newState })
      )
      return newState
    }
  }
}

// Exports class singleton to prevent multiple invocations of acquireVsCodeApi.
export const vscode = new VSCodeAPIWrapper()

/*
  Substantial portions of this code were taken from vscode-webview-ui-toolkit-samples
  https://github.com/microsoft/vscode-webview-ui-toolkit-samples/blob/6131aa6b49a49434003529e7b0fb7bff703dd02d/frameworks/hello-world-react-vite/webview-ui/src/utilities/vscode.ts

  MIT License

  Copyright (c) Microsoft Corporation.

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE
  */
