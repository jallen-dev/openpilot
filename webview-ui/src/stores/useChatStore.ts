import { create } from "zustand"
import { immer } from "zustand/middleware/immer"

import { ChatMessage } from "../../../src/shared/ChatMessage"
import { vscode } from "../utils/vscode"

interface ChatState {
  userPrompt: string
  loading: boolean // used to show loading indicator for non-streaming chats
  chatMessages: ChatMessage[]
  setUserPrompt: (prompt: string) => void
  setLoading: (loading: boolean) => void
  setChatMessages: (messages: ChatMessage[]) => void
}

export const useChatStore = create<ChatState>()(
  immer((set, get) => ({
    userPrompt: vscode.getState()?.userPrompt ?? "",
    chatMessages: vscode.getState()?.chatMessages ?? [],
    loading: false,
    setUserPrompt: (userPrompt: string) => {
      vscode.setState({ userPrompt })
      set({ userPrompt })
    },
    setLoading: (loading: boolean) => set({ loading }),
    setChatMessages: (chatMessages: ChatMessage[]) => {
      vscode.setState({ chatMessages })
      set({ chatMessages })
    }
  }))
)
