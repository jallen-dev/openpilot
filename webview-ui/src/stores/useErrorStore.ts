import { create } from "zustand"

import { vscode } from "../utils/vscode"

interface ErrorState {
  errorMessage: string
  setErrorMessage: (errorMessage: string) => void
}

export const useErrorStore = create<ErrorState>()((set, get) => ({
  errorMessage: vscode.getState()?.errorMessage ?? "",
  setErrorMessage: (errorMessage: string) => {
    vscode.setState({ errorMessage })
    set({ errorMessage })
  }
}))
