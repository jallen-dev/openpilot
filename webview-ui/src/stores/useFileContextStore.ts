import { create } from "zustand"

import { File, FileSearchResult } from "../../../src/shared/File"
import { vscode } from "../utils/vscode"

interface FileContextState {
  suggestedFiles: FileSearchResult[]
  rejectedFilePaths: string[]
  setSuggestedFiles: (files: FileSearchResult[]) => void
  rejectFile: (file: string) => void
  undoRejectFile: (file: string) => void
  clearRejectedFiles: () => void
  getFilesToInclude: () => File[]
}

export const useFileContextStore = create<FileContextState>()((set, get) => ({
  suggestedFiles: vscode.getState()?.suggestedFiles ?? [],
  rejectedFilePaths: vscode.getState()?.rejectedFilePaths ?? [],
  setSuggestedFiles: (suggestedFiles: FileSearchResult[]) => {
    vscode.setState({ suggestedFiles })
    set({ suggestedFiles })
  },
  rejectFile: (file: string) =>
    set((state) => {
      vscode.setState({ rejectedFilePaths: [...state.rejectedFilePaths, file] })
      return {
        rejectedFilePaths: [...state.rejectedFilePaths, file]
      }
    }),
  undoRejectFile: (file: string) =>
    set((state) => {
      vscode.setState({
        rejectedFilePaths: state.rejectedFilePaths.filter((f) => f !== file)
      })
      return {
        rejectedFilePaths: state.rejectedFilePaths.filter((f) => f !== file)
      }
    }),
  clearRejectedFiles: () => {
    vscode.setState({ rejectedFilePaths: [] })
    set({ rejectedFilePaths: [] })
  },
  getFilesToInclude: () => {
    const { suggestedFiles, rejectedFilePaths } = get()
    return suggestedFiles.filter((f) => !rejectedFilePaths.includes(f.path))
  }
}))
