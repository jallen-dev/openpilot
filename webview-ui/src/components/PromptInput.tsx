import { useCallback, useEffect, useRef } from "react"

import { ChatMessage } from "../../../src/shared/ChatMessage"
import {
  FileContentsMessage,
  MatchingFilesMessage
} from "../../../src/shared/OpenpilotMessage"
import useMessageListener from "../hooks/useMessageListener"
import { llmFactory } from "../services/LlmFactory"
import { useAppStateStore } from "../stores/useAppStateStore"
import { useChatStore } from "../stores/useChatStore"
import { useErrorStore } from "../stores/useErrorStore"
import { useSettingsStore } from "../stores/useSettingsStore"
import { vscode } from "../utils/vscode"

export default function PromptInput() {
  const userPrompt = useChatStore((state) => state.userPrompt)
  const setUserPrompt = useChatStore((state) => state.setUserPrompt)
  const loading = useChatStore((state) => state.loading)

  const apiKeys = useSettingsStore((state) => state.apiKeys)
  const includeFiles = useSettingsStore((state) => state.includeFiles)

  const textarea = useRef<HTMLTextAreaElement | null>(null)

  useEffect(() => {
    textarea.current!.style.height = textarea.current!.scrollHeight + "px"
  }, [])

  useMessageListener<MatchingFilesMessage>("matchingFiles", async (message) => {
    const candidateFilePaths = message.files.map((file) => file.path)
    const paths = await confirmFilePaths(candidateFilePaths)

    if (paths) {
      vscode.postMessage({
        type: "getFileContents",
        paths
      })
    }
  })

  useMessageListener<FileContentsMessage>("fileContents", async ({ files }) => {
    const model = useSettingsStore.getState().model
    const service = llmFactory(model)
    const chatMessages = useChatStore.getState().chatMessages
    await service.chat(chatMessages, files)
    useChatStore.getState().setLoading(false)
  })

  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLTextAreaElement>) => {
      const prompt = e.target.value
      setUserPrompt(prompt)

      // magic to make the textarea grow
      textarea.current!.style.height = "auto"
      textarea.current!.style.height =
        Math.min(150, textarea.current!.scrollHeight) + "px"
    },
    [includeFiles]
  )

  const hasKeys = apiKeys.OpenAI || apiKeys.Google
  const placeholder = hasKeys
    ? "Type your prompt and hit Enter. Shift+Enter to add a new line."
    : "Set your API keys in the settings."

  return (
    <textarea
      ref={textarea}
      className="w-full p-2 resize-none rounded-md"
      placeholder={placeholder}
      value={userPrompt}
      disabled={!hasKeys}
      onChange={handleChange}
      onKeyDown={(e) => {
        if (e.key === "Enter" && e.shiftKey === false) {
          e.preventDefault()

          if (hasKeys && !loading) {
            submitPrompt(userPrompt)
            setUserPrompt("")
          }
        }
      }}
    />
  )
}

async function submitPrompt(content: string) {
  useChatStore.getState().setLoading(true)
  useErrorStore.getState().setErrorMessage("")

  const newChat: ChatMessage = { content, participant: "user" }
  const chatMessages = [...useChatStore.getState().chatMessages, newChat]
  useChatStore.getState().setChatMessages(chatMessages)

  const model = useSettingsStore.getState().model
  const includeFiles = useSettingsStore.getState().includeFiles

  const workspaceIndexed = useAppStateStore.getState().workspaceIndexed

  const service = llmFactory(model)

  if (includeFiles && workspaceIndexed) {
    vscode.postMessage({
      type: "getFilesMatchingPrompt",
      prompt: content
    })
  } else {
    await service.chat(chatMessages)
    useChatStore.getState().setLoading(false)
  }
}

async function confirmFilePaths(candidateFileNames: string[]) {
  const chatMessages = useChatStore.getState().chatMessages
  if (chatMessages.length === 0) {
    return []
  }

  const model = useSettingsStore.getState().model
  const service = llmFactory(model)

  return service.askForContext(chatMessages, candidateFileNames)
}
