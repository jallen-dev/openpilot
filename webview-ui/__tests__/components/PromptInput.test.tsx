import { render, screen } from "@testing-library/react"

import PromptInput from "../../src/components/PromptInput"
import { useSettingsStore } from "../../src/stores/useSettingsStore"
import { vscode } from "../../src/utils/vscode"

const postMessage = vi.spyOn(vscode, "postMessage")
afterEach(() => {
  vi.resetAllMocks()
})

beforeAll(() => {
  vi.useFakeTimers({ shouldAdvanceTime: true })
})

test("textbox is disabled when there are no keys", () => {
  useSettingsStore.getState().setApiKeys({ OpenAI: "" })
  useSettingsStore.getState().setIncludeFiles(true)
  render(<PromptInput />)

  expect(screen.getByRole("textbox")).toBeDisabled()
})

test("placeholder is correct when there are no keys", () => {
  useSettingsStore.getState().setApiKeys({ OpenAI: "" })
  render(<PromptInput />)

  expect(screen.getByRole("textbox")).toHaveAttribute(
    "placeholder",
    "Set your API keys in the settings."
  )
})

test("placeholder is correct when there are keys", () => {
  useSettingsStore.getState().setApiKeys({ OpenAI: "test" })
  render(<PromptInput />)

  expect(screen.getByRole("textbox")).toHaveAttribute(
    "placeholder",
    "Type your prompt and hit Enter. Shift+Enter to add a new line."
  )
})

test("matches snapshot", () => {
  const { asFragment } = render(<PromptInput />)
  expect(asFragment()).toMatchSnapshot()
})
