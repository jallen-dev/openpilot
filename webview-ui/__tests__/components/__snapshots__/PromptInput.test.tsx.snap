// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <textarea
    class="w-full p-2 resize-none rounded-md"
    disabled=""
    placeholder="Set your API keys in the settings."
    style="height: 0px;"
  />
</DocumentFragment>
`;
